src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java

```diff
@@ -0,0 +1,201 @@
+package graphql.execution.reactive;
+
+import graphql.Internal;
+import graphql.util.LockKit;
+import org.jetbrains.annotations.NotNull;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * This subscriber can be used to map between a {@link org.reactivestreams.Publisher} of U
+ * elements and map them into {@link CompletionStage} of D promises.
+ *
+ * @param <U> published upstream elements
+ * @param <D> mapped downstream values
+ */
+@Internal
+public class CompletionStageSubscriber<U, D> implements Subscriber<U> {
+    protected final Function<U, CompletionStage<D>> mapper;
+    protected final Subscriber<? super D> downstreamSubscriber;
+    protected Subscription delegatingSubscription;
+    protected final Queue<CompletionStage<?>> inFlightDataQ;
+    protected final LockKit.ReentrantLock lock = new LockKit.ReentrantLock();
+    protected final AtomicReference<Runnable> onCompleteRun;
+    protected final AtomicBoolean isTerminal;
+
+    public CompletionStageSubscriber(Function<U, CompletionStage<D>> mapper, Subscriber<? super D> downstreamSubscriber) {
+        this.mapper = mapper;
+        this.downstreamSubscriber = downstreamSubscriber;
+        inFlightDataQ = new ArrayDeque<>();
```

Comment:
Nitpick: to a non-native English speaker it's going to be confusing what the "Q" is (it being a homophone for "queue"). I can see it's not introduced in this PR, it's based off naming from previous code.

Could rename these methods/variables to "Queue" or mention in a comment that the "Q" means queue.